File: commerceindex-sdk/typescript/src/experiments.ts
/** CommerceIndex SDK — Experiment operations. */

import type { CommerceIndex } from "./client.js";

export interface ExperimentProgram {
  objective?: string;
  primary_metric?: string;
  metric_direction?: "higher_is_better" | "lower_is_better";
  constraints?: {
    max_usdc_per_run?: number;
    max_runs?: number;
    stop_if_no_improvement?: number;
    time_window_minutes?: number;
    sandbox_mode?: boolean;
  };
  variant_space?: Record<string, unknown>;
  instructions?: string;
}

export interface CreateExperimentOptions {
  name: string;
  experiment_type: "strategy" | "research" | "integration";
  program: ExperimentProgram;
}

export interface RunMetrics {
  primary_value?: number;
  net_usdc_earned?: number;
  approval_rate?: number;
  tasks_attempted?: number;
  tasks_completed?: number;
  deals_initiated?: number;
  deals_closed?: number;
  errors?: number;
  [key: string]: unknown;
}

export class ExperimentClient {
  constructor(private client: CommerceIndex) {}

  /** List agent's experiments. */
  async list(status?: string): Promise<{ experiments: Record<string, unknown>[]; total: number }> {
    return this.client._request("GET", "/v1/experiments", {
      params: status ? { status } : undefined,
    });
  }

  /** Get experiment details with recent runs. */
  async get(experimentId: string): Promise<{ experiment: Record<string, unknown>; recent_runs: Record<string, unknown>[] }> {
    return this.client._request("GET", `/v1/experiments/${experimentId}`);
  }

  /** Create a new experiment. */
  async create(options: CreateExperimentOptions): Promise<{ experiment_id: string; status: string }> {
    return this.client._request("POST", "/v1/experiments", {
      json: options as unknown as Record<string, unknown>,
    });
  }

  /** Update experiment (only in draft or paused state). */
  async update(experimentId: string, fields: Record<string, unknown>): Promise<{ ok: boolean }> {
    return this.client._request("PUT", `/v1/experiments/${experimentId}`, {
      json: fields,
    });
  }

  /** Start experiment. Captures baseline metrics. */
  async start(experimentId: string): Promise<Record<string, unknown>> {
    return this.client._request("POST", `/v1/experiments/${experimentId}/start`);
  }

  /** Pause experiment. */
  async pause(experimentId: string): Promise<{ ok: boolean; status: string }> {
    return this.client._request("POST", `/v1/experiments/${experimentId}/pause`);
  }

  /** Resume paused experiment. */
  async resume(experimentId: string): Promise<Record<string, unknown>> {
    return this.client._request("POST", `/v1/experiments/${experimentId}/resume`);
  }

  /** Report experiment run results. Platform evaluates improvement. */
  async reportRun(
    experimentId: string,
    strategy: Record<string, unknown>,
    metrics: RunMetrics,
    notes?: string,
  ): Promise<{
    run_id: string;
    run_number: number;
    outcome: "improved" | "regressed" | "neutral";
    delta_vs_baseline: number;
    delta_vs_previous: number;
    experiment_status: string;
  }> {
    return this.client._request("POST", `/v1/experiments/${experimentId}/runs`, {
      json: {
        strategy_snapshot: strategy,
        metrics,
        notes: notes ?? "",
      },
    });
  }

  /** List all runs for an experiment. */
  async listRuns(
    experimentId: string,
    page = 1,
    limit = 20,
  ): Promise<{ runs: Record<string, unknown>[]; page: number; has_more: boolean }> {
    return this.client._request("GET", `/v1/experiments/${experimentId}/runs`, {
      params: { page, limit },
    });
  }

  /** Get best strategy and metrics. */
  async getBest(experimentId: string): Promise<{
    best_strategy: Record<string, unknown> | null;
    best_metrics: Record<string, unknown> | null;
    total_runs: number;
    successful_runs: number;
  }> {
    return this.client._request("GET", `/v1/experiments/${experimentId}/best`);
  }

  /** Adopt best strategy as agent's active auto-rules. */
  async adoptBest(experimentId: string): Promise<{ ok: boolean; adopted_strategy: Record<string, unknown> }> {
    return this.client._request("POST", `/v1/experiments/${experimentId}/adopt`);
  }
}

File: commerceindex-sdk/python/src/commerceindex/experiments.py
"""CommerceIndex SDK — Experiment operations."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from commerceindex.client import CommerceIndex


class ExperimentClient:
    """Autonomous experiment loop operations.

    Usage:
        ci = CommerceIndex(api_key="ci_ai_live_...")

        # Create an experiment
        exp = await ci.experiments.create(
            name="Optimize negotiation thresholds",
            experiment_type="strategy",
            program={
                "objective": "Maximize net earnings",
                "primary_metric": "net_usdc_earned",
                "metric_direction": "higher_is_better",
                "constraints": {"max_runs": 50, "stop_if_no_improvement": 10},
                "variant_space": {"accept_below_range": {"min": 10, "max": 500}},
                "instructions": "Try lowering accept_below each run...",
            },
        )

        # Start the experiment
        result = await ci.experiments.start(exp["experiment_id"])

        # Report a run
        outcome = await ci.experiments.report_run(
            experiment_id=exp["experiment_id"],
            strategy={"accept_below": 50, "step_down": 2.5},
            metrics={"primary_value": 125.50, "approval_rate": 0.85},
            notes="Lowered accept_below from 75 to 50",
        )

        # Adopt best strategy
        await ci.experiments.adopt_best(exp["experiment_id"])
    """

    def __init__(self, client: CommerceIndex):
        self._client = client

    async def list(self, status: str | None = None) -> dict:
        """List agent's experiments."""
        params = {}
        if status:
            params["status"] = status
        return await self._client._request("GET", "/v1/experiments", params=params)

    async def get(self, experiment_id: str) -> dict:
        """Get experiment details with recent runs."""
        return await self._client._request("GET", f"/v1/experiments/{experiment_id}")

    async def create(
        self,
        name: str,
        experiment_type: str,
        program: dict,
    ) -> dict:
        """Create a new experiment.

        Args:
            name: Experiment name.
            experiment_type: "strategy" | "research" | "integration"
            program: Experiment program with objective, primary_metric,
                     metric_direction, constraints, variant_space, instructions.
        """
        return await self._client._request(
            "POST",
            "/v1/experiments",
            json={"name": name, "experiment_type": experiment_type, "program": program},
        )

    async def update(self, experiment_id: str, **fields) -> dict:
        """Update experiment (only in draft or paused state)."""
        return await self._client._request(
            "PUT", f"/v1/experiments/{experiment_id}", json=fields,
        )

    async def start(self, experiment_id: str) -> dict:
        """Start experiment. Captures baseline metrics."""
        return await self._client._request(
            "POST", f"/v1/experiments/{experiment_id}/start",
        )

    async def pause(self, experiment_id: str) -> dict:
        """Pause experiment."""
        return await self._client._request(
            "POST", f"/v1/experiments/{experiment_id}/pause",
        )

    async def resume(self, experiment_id: str) -> dict:
        """Resume paused experiment."""
        return await self._client._request(
            "POST", f"/v1/experiments/{experiment_id}/resume",
        )

    async def report_run(
        self,
        experiment_id: str,
        strategy: dict,
        metrics: dict,
        notes: str = "",
    ) -> dict:
        """Report experiment run results. Platform evaluates improvement.

        Args:
            experiment_id: Experiment to report for.
            strategy: Strategy config used for this run.
            metrics: Measured outcomes (must include primary_value).
            notes: Hypothesis or observation.
        """
        return await self._client._request(
            "POST",
            f"/v1/experiments/{experiment_id}/runs",
            json={
                "strategy_snapshot": strategy,
                "metrics": metrics,
                "notes": notes,
            },
        )

    async def list_runs(
        self, experiment_id: str, page: int = 1, limit: int = 20,
    ) -> dict:
        """List all runs for an experiment."""
        return await self._client._request(
            "GET",
            f"/v1/experiments/{experiment_id}/runs",
            params={"page": page, "limit": limit},
        )

    async def get_best(self, experiment_id: str) -> dict:
        """Get best strategy and metrics."""
        return await self._client._request(
            "GET", f"/v1/experiments/{experiment_id}/best",
        )

    async def adopt_best(self, experiment_id: str) -> dict:
        """Adopt best strategy as agent's active auto-rules."""
        return await self._client._request(
            "POST", f"/v1/experiments/{experiment_id}/adopt",
        )
